Add profile toggle value to optionally show your e-mail address to authenticated users

joancodes opened the pull request:
# Notable Changes

* This PR aims to add the show_email boolean attribute to user settings for users to toggle if they want to publicise their email addresses to other authenticated users.

## Screenshots (Optional)

<img width="965" alt="profile-with-email Screenshot " src="https://github.com/lobsters/lobsters/assets/25585114/cb5115e2-259a-40ee-99db-c19e6cfad82e">

<img width="706" alt="show-email-setting Screenshot" src="https://github.com/lobsters/lobsters/assets/25585114/33530f72-0229-4197-9c62-5e5e8ee059cd">


## Todo

* N/A


<!--
Issues and PRs are typically reviewed Wednesday and most weekend mornings.
-->

pushcx commented:
Mentioning #1207 so GitHub links them

pushcx commented on app/views/settings/index.html.erb:
```diff
@@ -181,6 +181,14 @@
 
     <h2>Miscellaneous Settings</h2>
 
```
Please remove this `if`, all users should be able to check this box.

pushcx commented on db/schema.rb:
```diff
@@ -222,7 +222,7 @@
     t.text "description", size: :medium
     t.text "body", size: :medium
-    t.timestamp "created_at", default: -> { "current_timestamp() ON UPDATE current_timestamp()" }, null: false
+    t.timestamp "created_at", default: -> { "CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP" }, null: false
```
Please remove this unrelated change.

pushcx commented on db/schema.rb:
```diff
@@ -365,6 +366,6 @@
 
   create_view "replying_comments", sql_definition: <<-SQL
-      select `read_ribbons`.`user_id` AS `user_id`,`comments`.`id` AS `comment_id`,`read_ribbons`.`story_id` AS `story_id`,`comments`.`parent_comment_id` AS `parent_comment_id`,`comments`.`created_at` AS `comment_created_at`,`parent_comments`.`user_id` AS `parent_comment_author_id`,`comments`.`user_id` AS `comment_author_id`,`stories`.`user_id` AS `story_author_id`,`read_ribbons`.`updated_at` < ` ...
+      select `read_ribbons`.`user_id` AS `user_id`,`comments`.`id` AS `comment_id`,`read_ribbons`.`story_id` AS `story_id`,`comments`.`parent_comment_id` AS `parent_comment_id`,`comments`.`created_at` AS `comment_created_at`,`parent_comments`.`user_id` AS `parent_comment_author_id`,`comments`.`user_id` AS `comment_author_id`,`stories`.`user_id` AS `story_author_id`,(`read_ribbons`.`updated_at` <  ...
```
Please remove this unrelated change.

joancodes committed:
Remove unnecessary conditional  and schema changes

joancodes commented:
Hi @pushcx, I have made the necessary changes.

epilys reviewed (changes requested):
Does this show the e-mail on authenticated users only? I can't find a check anywhere.

Also the option could probably use a label that clarifies this won't be completely public.

joancodes commented on app/views/users/show.html.erb:
```diff
@@ -100,6 +100,14 @@
     <div style="clear: both;"></div>
   <% end %>
 
+  <% if @user.is_admin? && @user.show_email? %>
```
Hi @pushcx, could this be the check you are referring to?

joancodes replied:
Also, I am up for suggestions on what text to use on the label.

pushcx replied:
I was referring to the setting page. This `if` should be `||` so admins always see emails.

pushcx commented on app/views/settings/index.html.erb:
```diff
@@ -181,6 +181,12 @@
+    <div class="boxline">
+      <%= f.label :show_email, "Show Email Address:",
+        :class => "required" %>
+      <%= f.check_box :show_email %>
```
Please add a `<span class=“hint indent”>` like on Username above, with the text “Only shown to logged-in users”

joancodes committed:
feature/add-email-visibilty-toggle: Updated email visibility settings and conditions in UI

Added hint for email visibility to logged-in users and adjusted conditions for email display to users who have opted to show their email.

joancodes commented:
Hi @pushcx, I have added some more changes. LMK if I have missed anything.

pushcx commented:
@joancodes Thanks for contributing this feature and being patient with changes. I’m really glad to merge this today.

joancodes commented:
> 🎉 kudos @joancodes I just enabled it on lobste.rs and it works!

Glad to hear this! Thank you for letting me work on this and looking forward to making more meaningful contributions to lobste.rs 🎉

banteg commented:
this pr has introduced a privacy vulnerability that leaks every user's email.

any logged-in user could enable “show email” on their own account, visit another user’s profile, and see that other user’s email even if the other user did not opt in.

originally reported here https://lobste.rs/s/7heurd/kyaaa_your_emails_are_showing_lobste_rs

fixed here https://github.com/lobsters/lobsters/commit/7037494eb85003603ecd87fb00bde994e111f103
